ChopChop/Storage/StorageManager.swift

```diff
@@ -228,6 +233,172 @@ extension StorageManager {
     }
 }
 
+// MARK: - Firebase operations
+extension StorageManager {
+
+    func publishRecipe(recipe: inout Recipe, userId: String) throws {
+        var cuisine: String?
+        if let categoryId = recipe.recipeCategoryId {
+            cuisine = try? fetchRecipeCategory(id: categoryId)?.name
+        }
+        let steps = recipe.steps.map({ $0.content })
+        let ingredients = recipe.ingredients.map({
+            OnlineIngredientRecord(name: $0.name, quantity: $0.quantity.record)
+        })
+        let recipeRecord = OnlineRecipeRecord(
+            name: recipe.name,
+            creator: userId,
+            servings: recipe.servings,
+            cuisine: cuisine,
+            difficulty: recipe.difficulty,
+            ingredients: ingredients,
+            steps: steps
+        )
+        let onlineId = try firebase.addRecipe(recipe: recipeRecord)
+        recipe.onlineId = onlineId
+        try self.saveRecipe(&recipe)
+    }
+
+    // this should only be called once when the app first launched
+    func createUser(name: String) throws -> String {
+        let user = User(name: name)
+        return try firebase.addUser(user: user)
+    }
+
+    func updateOnlineRecipe(recipe: Recipe, userId: String) throws {
+        var cuisine = ""
+        if let categoryId = recipe.recipeCategoryId {
+            cuisine = (try? (fetchRecipeCategory(id: categoryId)?.name) ?? "") ?? ""
+        }
+        let steps = recipe.steps.map({ $0.content })
+        let ingredients = recipe.ingredients.map({
+            OnlineIngredientRecord(name: $0.name, quantity: $0.quantity.record)
+        })
+        let recipeRecord = OnlineRecipeRecord(
+            id: recipe.onlineId,
+            name: recipe.name,
+            creator: userId,
+            servings: recipe.servings,
+            cuisine: cuisine,
+            difficulty: recipe.difficulty,
+            ingredients: ingredients,
+            steps: steps
+        )
+        try firebase.updateRecipeDetails(recipe: recipeRecord)
+    }
+
+    func removeRecipeFromOnline(recipe: OnlineRecipe) throws {
+        try firebase.removeRecipe(recipeId: recipe.id)
+        for rating in recipe.ratings {
+            firebase.removeUserRecipeRating(userId: rating.userId, rating: UserRating(recipeOnlineId: recipe.id, score: rating.score))
+        }
+    }
+
+    func fetchOnlineRecipeById(recipeId: String) -> AnyPublisher<OnlineRecipe, Error> {
```

Comment:
can check if this is able to observe changes in arrays? ie the ratings array?

Reply:
yes supposed to